pkg/controllers/statusaggregator/plugins/job.go

```diff
@@ -0,0 +1,163 @@
+)
+
+type JobPlugin struct{}
+
+func NewJobPlugin() *JobPlugin {
+	return &JobPlugin{}
+}
+
+func (receiver *JobPlugin) AggregateStatuses(
+	ctx context.Context,
+	sourceObject, fedObject *unstructured.Unstructured,
+	clusterObjs map[string]interface{},
+	clusterObjsUpToDate bool,
+) (*unstructured.Unstructured, bool, error) {
+	logger := klog.FromContext(ctx).WithValues("status-aggregator-plugin", "jobs")
+
+	needUpdate := false
+	aggregatedStatus := &batchv1.JobStatus{}
+	var completionTime *metav1.Time
+	var completionCount int
+	var jobCompleted, jobFailed []string
+
+	for clusterName, clusterObj := range clusterObjs {
+		logger := klog.FromContext(ctx).WithValues("cluster-name", clusterName)
+
+		utd := clusterObj.(*unstructured.Unstructured)
+		// For status of job
+		var found bool
+		status, found, err := unstructured.NestedMap(utd.Object, common.StatusField)
+		if err != nil || !found {
+			logger.Error(err, "Failed to get status of cluster object")
+			return nil, false, err
+		}
+
+		if status == nil {
+			continue
+		}
+
+		jobStatus := &batchv1.JobStatus{}
+		if err = runtime.DefaultUnstructuredConverter.FromUnstructured(status, jobStatus); err != nil {
+			logger.Error(err, "Failed to convert the status of cluster object")
+			return nil, false, err
+		}
+
+		if aggregatedStatus.StartTime.IsZero() || jobStatus.StartTime.Before(aggregatedStatus.StartTime) {
+			aggregatedStatus.StartTime = jobStatus.StartTime
+		}
+		if !jobStatus.CompletionTime.IsZero() {
+			completionCount++
+			jobCompleted = append(jobCompleted, clusterName)
+			if completionTime.IsZero() || completionTime.Before(jobStatus.CompletionTime) {
+				completionTime = jobStatus.CompletionTime
+			}
+		} else if IsJobFinishedWithFailed(jobStatus) {
+			completionCount++
+			jobFailed = append(jobFailed, clusterName)
+		}
+
+		aggregatedStatus.Active += jobStatus.Active
+		aggregatedStatus.Succeeded += jobStatus.Succeeded
+		aggregatedStatus.Failed += jobStatus.Failed
+	}
+	if completionCount > 0 && completionCount == len(clusterObjs) {
+		now := time.Now()
+		var conditionType batchv1.JobConditionType
+		var reason, message string
+		sort.Strings(jobCompleted)
+		sort.Strings(jobFailed)
+
+		switch {
+		case len(jobCompleted) > 0 && len(jobFailed) > 0:
+			conditionType = batchv1.JobFailed
+			reason = "Mixed"
+			message = fmt.Sprintf("Job completed in clusters [%s] and failed in member clusters [%s]",
+				strings.Join(jobCompleted, ","), strings.Join(jobFailed, ","))
+		case len(jobCompleted) > 0:
+			conditionType = batchv1.JobComplete
+			reason = "Completed"
+			message = fmt.Sprintf("Job completed in clusters [%s]", strings.Join(jobCompleted, ","))
+		default:
+			conditionType = batchv1.JobFailed
+			reason = "Failed"
+			message = fmt.Sprintf("Job failed in clusters [%s]", strings.Join(jobFailed, ","))
+		}
+
+		aggregatedStatus.CompletionTime = completionTime
+		aggregatedStatus.Conditions = append(aggregatedStatus.Conditions, batchv1.JobCondition{
```

Comment:
Would this append the same `Failed` or `Completed` conditions infinitely during successive reconciles? We should either overwrite the old conditions here, or prevent updating the status if the job is already finished. @mrlihanbo @limhawjia Which one sounds more reasonable?

Reply:
I think we should try to follow the semantics of the job controller as much as possible, which seems to be to skip updating finished jobs

Reply:
After giving it some thought, we should continue to update it so as to not lose information about the remaining clusters' jobs after the host job is marked as `Failed`. 

@gary-lgy there shouldn't be an infinite loop as the FederatedJobStatus itself should not be a reconcile trigger?

Reply:
Can we solve the infinite reconcile loop problem by checking the type, status, reason and message? If they are all equal, we skip updating.